mm2src/coins/hd_pubkey.rs

```diff
@@ -0,0 +1,244 @@
+        }
+    }
+}
+
+#[async_trait]
+pub trait ExtractExtendedPubkey {
+    type ExtendedPublicKey;
+
+    async fn extract_extended_pubkey<XPubExtractor>(
+        &self,
+        xpub_extractor: &XPubExtractor,
+        derivation_path: DerivationPath,
+    ) -> MmResult<Self::ExtendedPublicKey, HDExtractPubkeyError>
+    where
+        XPubExtractor: HDXPubExtractor + Sync;
+}
+
+#[async_trait]
+pub trait HDXPubExtractor {
+    async fn extract_utxo_xpub(
+        &self,
+        trezor_utxo_coin: TrezorUtxoCoin,
+        derivation_path: DerivationPath,
+    ) -> MmResult<XPub, HDExtractPubkeyError>;
+}
+
+pub enum RpcTaskXPubExtractor<'task, Item, Error, InProgressStatus, AwaitingStatus, UserAction>
+where
+    Item: Serialize,
+    Error: SerMmErrorType,
+{
+    Trezor {
+        hw_ctx: HardwareWalletArc,
+        task_handle: &'task RpcTaskHandle<Item, Error, InProgressStatus, AwaitingStatus, UserAction>,
+        statuses: HwConnectStatuses<InProgressStatus, AwaitingStatus>,
+    },
+}
+
+#[async_trait]
+impl<'task, Item, Error, InProgressStatus, AwaitingStatus, UserAction> HDXPubExtractor
+    for RpcTaskXPubExtractor<'task, Item, Error, InProgressStatus, AwaitingStatus, UserAction>
+where
+    Item: Serialize + Send,
+    Error: SerMmErrorType + Send,
+    InProgressStatus: Clone + Send + Sync,
+    AwaitingStatus: Clone + Send + Sync,
+    UserAction: TryInto<TrezorPinMatrix3x3Response, Error = RpcTaskError> + Send,
+{
+    async fn extract_utxo_xpub(
+        &self,
+        trezor_utxo_coin: TrezorUtxoCoin,
+        derivation_path: DerivationPath,
+    ) -> MmResult<XPub, HDExtractPubkeyError> {
+        match self {
+            RpcTaskXPubExtractor::Trezor {
+                hw_ctx,
+                task_handle,
+                statuses,
+            } => {
+                Self::extract_utxo_xpub_from_trezor(hw_ctx, task_handle, statuses, trezor_utxo_coin, derivation_path)
+                    .await
+            },
+        }
+    }
+}
+
+impl<'task, Item, Error, InProgressStatus, AwaitingStatus, UserAction>
+    RpcTaskXPubExtractor<'task, Item, Error, InProgressStatus, AwaitingStatus, UserAction>
+where
+    Item: Serialize,
+    Error: SerMmErrorType,
+{
+    pub fn new(
+        ctx: &MmArc,
+        task_handle: &'task RpcTaskHandle<Item, Error, InProgressStatus, AwaitingStatus, UserAction>,
+        statuses: HwConnectStatuses<InProgressStatus, AwaitingStatus>,
+    ) -> MmResult<Self, HDExtractPubkeyError> {
+        let crypto_ctx = CryptoCtx::from_ctx(ctx)?;
+        // Don't use [`CryptoCtx::hw_ctx`] because we are planning to support HD master key.
+        match *crypto_ctx {
+            CryptoCtx::HardwareWallet(ref hw_ctx) => Ok(RpcTaskXPubExtractor::Trezor {
+                hw_ctx: hw_ctx.clone(),
+                task_handle,
+                statuses,
+            }),
+            CryptoCtx::KeyPair(_) => MmError::err(HDExtractPubkeyError::IguanaPrivKeyNotAllowed),
```

Comment:
Can we avoid these repeated checks?

Reply:
Will try to avoid

Reply:
I hope I did it :)

Reply:
Could you please clarify what was changed? 🙂